textworld/generator/chaining.py

```diff
@@ -296,3 +296,211 @@ def print_chains(chains, verbose=False, backward=False):
+
+            for j, action in enumerate(actions):
+                state = self.apply(node.state, action)
+                if not state:
+                    continue
+
+                remaining = actions[j+1:]
+                new_backtracks = backtracks + [remaining]
+                yield _ChainNode(node, state, action, new_backtracks, i + 1, node.breadth + 1)
+
+    def is_relevant(self, node, action):
+        if not node.action:
+            return True
+
+        if self.backward:
+            # XXX
+            post = action.removed
+            pre = node.action.postconditions
+        else:
+            post = node.action.added
+            pre = action.preconditions
+
+        return bool(set(post) & set(pre))
+
+    def apply(self, state, action):
+        new_state = state.copy()
+        for prop in action.preconditions:
+            if not new_state.is_fact(prop):
+                if all(state.has_variable(var) for var in prop.arguments):
+                    # Don't allow creating new predicates without any new variables
+                    return None
+                new_state.add_fact(prop)
+
+        # Make sure new_state still respects the constraints
+        if not self.check_state(new_state):
+            return None
+
+        new_state.apply(action)
+
+        # XXX: Some debug checks
+        assert self.check_state(new_state)
+
+        return new_state
+
+    def check_state(self, state):
+        fail = Proposition("fail", [])
+
+        constraints = state.all_applicable_actions(self.constraints)
+        for constraint in constraints:
+            if state.is_applicable(constraint):
+                # Optimistically delay copying the state
+                copy = state.copy()
+                copy.apply(constraint)
+
+                if copy.is_fact(fail):
+                    return False
+
+        return True
+
+    def try_instantiate(self, state, rule, mapping):
+        for ph in rule.placeholders:
+            if mapping.get(ph) is None:
+                # XXX: types_counts, max_types_counts
+                name = get_new(ph.type)
+                mapping[ph] = Variable(name, ph.type)
+
+        return rule.instantiate(mapping)
+
+    def all_assignments(self, state, rules):
+        assignments = []
+        for rule in rules:
+            if self.backward:
+                rule = rule.inverse()
+            # XXX: self.constants_mapping?
+            # XXX: constrained_types
+            #for mapping in state.all_assignments(rule, data.get_types().constants_mapping, self.create_variables):
+            for mapping in state.all_assignments(rule, {}, self.create_variables):
+                assignments.append((rule, mapping))
+
+        # Keep everything in a deterministic order
+        return sorted(assignments, key=_assignment_sort_key)
+
+    def is_complete_chain(self, node):
+        return node.depth >= self.min_depth
+
+    def make_quest(self, node):
```

Comment:
I think we need to keep track of the initial state too. Or maybe I missed how we can get it?

Reply:
Yep I haven't really thought about the ideal return type here.  Something that combines a `Quest` and an initial `State`?

Reply:
I think just the list of `Action` and the initial `State` would be fine. I'm not sure if it makes sense for the `chain` function to return a `Quest`. 

Also, maybe we only need the list of `Action` from which we can infer the missing `Proposition` to add to an existing `State` object? That would allow us to easily generate different games sharing the same goal without the cost of doing chaining.

Reply:
> I think just the list of Action and the initial State would be fine. I'm not sure if it makes sense for the chain function to return a Quest.

Okay sounds good.

> Also, maybe we only need the list of Action from which we can infer the missing Proposition to add to an existing State object? That would allow us to easily generate different games sharing the same goal without the cost of doing chaining.

Sorry, not sure what you mean by this.

Reply:
I meant having a function that takes a `State` and a list of `Action`s and returns a new `State` for which the actions can be applied. Essentially, doing what the chaining (with `create_variable=True`) does but with a fixed list of actions. I guess that could easily be done once we have the callbacks replacing `rules_per_depth` so we can force the chaining to use the actions we want.